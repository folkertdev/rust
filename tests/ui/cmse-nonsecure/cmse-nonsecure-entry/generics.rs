//@ compile-flags: --target thumbv8m.main-none-eabi --crate-type lib
//@ needs-llvm-components: arm
#![feature(cmse_nonsecure_entry, no_core, lang_items)]
#![no_core]
#[lang = "sized"]
pub trait Sized {}
#[lang = "copy"]
pub trait Copy {}
impl Copy for u32 {}

#[repr(C)]
struct Wrapper<T>(T);

impl<T: Copy> Wrapper<T> {
    extern "C-cmse-nonsecure-entry" fn ambient_generic(_: T, _: u32, _: u32, _: u32) -> u64 {
        //~^ ERROR [E0798]
        0
    }

    extern "C-cmse-nonsecure-entry" fn ambient_generic_nested(
        //~^ ERROR [E0798]
        _: Wrapper<T>,
        _: u32,
        _: u32,
        _: u32,
    ) -> u64 {
        0
    }
}

extern "C-cmse-nonsecure-entry" fn introduced_generic<U: Copy>(
    //~^ ERROR [E0798]
    _: U,
    _: u32,
    _: u32,
    _: u32,
) -> u64 {
    0
}

extern "C-cmse-nonsecure-entry" fn impl_trait(_: impl Copy, _: u32, _: u32, _: u32) -> u64 {
    //~^ ERROR [E0798]
    0
}
