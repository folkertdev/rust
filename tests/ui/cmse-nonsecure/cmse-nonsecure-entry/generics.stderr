error[E0798]: functions with the `"C-cmse-nonsecure-entry"` ABI cannot contain generics in their type
  --> $DIR/generics.rs:31:1
   |
LL | / extern "C-cmse-nonsecure-entry" fn introduced_generic<U: Copy>(
LL | |
LL | |     _: U,
LL | |     _: u32,
LL | |     _: u32,
LL | |     _: u32,
LL | | ) -> u64 {
   | |________^

error[E0798]: functions with the `"C-cmse-nonsecure-entry"` ABI cannot contain generics in their type
  --> $DIR/generics.rs:41:1
   |
LL | extern "C-cmse-nonsecure-entry" fn impl_trait(_: impl Copy, _: u32, _: u32, _: u32) -> u64 {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0798]: functions with the `"C-cmse-nonsecure-entry"` ABI cannot contain generics in their type
  --> $DIR/generics.rs:15:5
   |
LL |     extern "C-cmse-nonsecure-entry" fn ambient_generic(_: T, _: u32, _: u32, _: u32) -> u64 {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0798]: functions with the `"C-cmse-nonsecure-entry"` ABI cannot contain generics in their type
  --> $DIR/generics.rs:20:5
   |
LL | /     extern "C-cmse-nonsecure-entry" fn ambient_generic_nested(
LL | |
LL | |         _: Wrapper<T>,
LL | |         _: u32,
LL | |         _: u32,
LL | |         _: u32,
LL | |     ) -> u64 {
   | |____________^

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0798`.
